### ISSUE TITLE
Debugger not working with web3, fails after Debugger.forTx call

### ISSUE THREAD

**Issue by davidcallanan** [trufflesuite/truffle#4635]
- [ ] I've [opened a support ticket](https://trufflesuite.zendesk.com/hc/en-us/requests/new) before filing this issue.

---------------------------

## Issue

I am trying to use the `@truffle/debugger` package with a web3 provider.

There is an error after calling `await Debugger.forTx(...)`

## Steps to Reproduce

```
import Web3 from "web3";
import Debugger from "@truffle/debugger";

let web3provider = new Web3.providers.HttpProvider("https://eth-mainnet.alchemyapi.io/v2/my_secret_key");
let web3 = new Web3(web3provider, null, {
	transactionConfirmationBlocks: 1,
});

let contracts = [... generated programatically ...];
let transactionHash = "0x03e784bfcb8f6175f7a321586f671dff08a87070f62f5c9ae928e8f2c31e17bd"; // randomly chosen from etherscan

// error happens during this function call:
let bugger = await Debugger.forTx(
	transactionHash,
	{
		provider: web3,
		contracts,
	},
);
```

## Expected Behavior

Debugging should not error.

## Actual Results

```
C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\web3-providers-http@1.7.0\node_modules\web3-providers-http\lib\index.js:86
    var request = this._prepareRequest();
                       ^
TypeError: this._prepareRequest is not a function
    at HttpProvider.send (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\web3-providers-http@1.7.0\node_modules\web3-providers-http\lib\index.js:86:24)
    at node:internal/util:360:7
    at new Promise (<anonymous>)
    at promisify (node:internal/util:346:12)
    at xn.getTrace (C:\Users\leefaria\code\truffle\packages\debugger\lib\web3\adapter.js:13:24)
    at runCallEffect (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:528:21)
    at runEffect (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:1208:7)
    at digestEffect (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:1275:5)
    at next (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:1165:9)
    at proc (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:1112:3)
    at C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:589:17
    at immediately (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:60:12)
    at runForkEffect (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:588:3)
    at runEffect (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:1208:7)
    at digestEffect (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:1275:5)
    at next (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\@redux-saga+core@1.1.3\node_modules\@redux-saga\core\dist\redux-saga-core.dev.cjs.js:1165:9)
```

Although this error is originating from web3, I feel that this may be a problem with this package, since I don't generally get random web3 errors, and this package is supposed to work with web3.

## Environment

* Operating System: Windows 10
* Ethereum client: an alchemy rpc node
* Truffle version (`truffle version`): `"@truffle/debugger": "^9.2.12"`
* node version (`node --version`): `v17.2.0`
* npm version (`npm --version`): `8.1.4`

**Comment by haltman-at** [comment]
Yes, in web3.js, there's a difference between the `web3` object and the underlying provider.  (This is different from how Ethers works; it collapses this distinction.)  Rather than passing in `web3`, you should pass in `web3.currentProvider`.  Hopefully that should fix your problem.  Let us know if you have any further problems!

**Comment by davidcallanan** [comment]
Thank you, that makes total sense.

However, I am now getting the following error:

```
var request = this._prepareRequest();
                       ^
TypeError: this._prepareRequest is not a function
    at HttpProvider.send (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\web3-providers-http@1.7.0\node_modules\web3-providers-http\lib\index.js:86:24)
```

I think I have everything right except one thing: I haven't been able to figure out what to put in for the `ast` for the contracts. So maybe this is what's causing the problem. Can you explain to me what `ast` should be? I can't find any documentation on it.

**Comment by haltman-at** [comment]
Well, you can see here that the problem is coming from `Abi` and `abi-utils`, so it's more likely that the problem is the omission of the ABIs, not the ASTs.  The debugger can function without the ASTs (though with much reduced functionality).  I suppose notionally it ought to be able to function without the ABIs as well, but missing ABIs are not a case that's ever come up before, so that's not something we ever bothered to make it robust against.

As for the ASTs, these are the abstract syntax trees that the Solidity compiler outputs.  If you're using Truffle, you can find them in the `ast` field in your Truffle artifacts.  If not, well, it's the `ast` output from solc.  You can read more [here](https://docs.soliditylang.org/en/v0.8.11/using-the-compiler.html#compiler-input-and-output-json-description) to see how to request it from solc and obtain it from the output.

**Comment by davidcallanan** [comment]
Thank you again. This error is gone after adding the abi (which is missing from [the documentation](https://github.com/trufflesuite/truffle/tree/develop/packages/debugger)).

But now I am getting this problem:

```
    var request = this._prepareRequest();
                       ^
TypeError: this._prepareRequest is not a function
    at HttpProvider.send (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\web3-providers-http@1.7.0\node_modules\web3-providers-http\lib\index.js:86:24)
```

Regarding the `ast`, I can't find any output from `solc` even after including it in the `outputSelection`. I'll try to use truffle for this instead.

**Comment by haltman-at** [comment]
> Thank you again. This error is gone after adding the abi (which is missing from the documentation).

Oh, wow, I didn't even realize we had that documentation.  Yeah, that's way out of date, sorry.

> Regarding the ast, I can't find any output from solc even after including it in the outputSelection. I'll try to use truffle for this instead.

Remember, the ast is per-source, not per-contract, so it has to go in the sources part of the outputSelection, not the contracts part.  As described there, you use `""` as the contract name to specify the per-source information to get.

> But now I am getting this problem:

Isn't that the same problem you reported originally?  I'm confused.

**Comment by davidcallanan** [comment]
> Remember, the ast is per-source, not per-contract, so it has to go in the sources part of the outputSelection, not the contracts part. As described there, you use `""` as the contract name to specify the per-source information to get.

Sorry, I had it commented out the whole time 🤦‍♂️ I have the ast in place now.

> Isn't that the same problem you reported originally? I'm confused.

Yes... which is very confusing.

I just checked and the free node I am using does not support debugging, so maybe that's the problem. I'll see if I can get a node that supports debugging.

**Comment by haltman-at** [comment]
> I just checked and the free node I am using does not support debugging, so maybe that's the problem. I'll see if I can get a node that supports debugging.

That will indeed prevent the debugger from working, but it shouldn't lead to the particular problem you saw.  If you have more detailed reproduction steps I can try to reproduce this.

Note that if you want a node with debugging, you may be able to accomplish that by using Ganache to fork your node. :)  I'm not entirely clear on under what conditions that will or will not work, though.

**Comment by davidcallanan** [comment]
I actually was using Ganache's forking feature initially (after getting the email about it yesterday), but that was causing another error:

```
C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\ganache@7.0.0\node_modules\ganache\dist\node\webpack:\Ganache\chains\ethereum\ethereum\lib\src\provider.js:10
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose 
class did not declare it");
                                                                                             ^
TypeError: Cannot read private member from an object whose class did not declare it
    at __classPrivateFieldGet (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\ganache@7.0.0\node_modules\ganache\dist\node\webpack:\Ganache\chains\ethereum\ethereum\lib\src\provider.js:10:94)
    at send (C:\Users\David\Documents\git\crypto-debug\backend\node_modules\.pnpm\ganache@7.0.0\node_modules\ganache\dist\node\webpack:\Ganache\chains\ethereum\ethereum\lib\src\provider.js:223:16)
```

I tried switch to another node (getblock.io) which offers transaction tracing, still the same error.

Here's my code in full to reproduce, thanks :)

`app.js`:

```
import ganache from "ganache";
import Web3 from "web3";
import fs from "fs";
import solc from "solc";
import Debugger from "@truffle/debugger";

let ganache_provider = ganache.provider({
	fork: {
		network: "mainnet",
		// blockNumber: 12345678,
	},
	logging: {
		quiet: true,
	},
	chain: {
		vmErrorsOnRPCResponse: false,
	},
	// mining: {
	// 	blockTime: 10,
	// }
});

// let web3provider = ganache_provider;
let web3provider = new Web3.providers.HttpProvider("https://eth.getblock.io/mainnet/?api_key=<ADD_IN_API_KEY_HERE>");

let web3 = new Web3(web3provider, null, {
	transactionConfirmationBlocks: 1,
});

let contracts = [];

let promise_resolve;
let promise = new Promise((res) => {
	promise_resolve = res;
});

console.log("loading solidity compiler...");
solc.loadRemoteVersion("v0.7.4+commit.3f05b770", (error, solcSnapshot) => {
	console.log("starting compilation process...");

	if (error) {
		console.log("Error with loading remote solidity compiler!");
		console.error(error);
	}

	let source = fs.readFileSync("./code.sol").toString();

	let input = {
		language: "Solidity",
		sources: {
			"code.sol": {
				content: source,
			},
		},
		settings: {
			outputSelection: {
				'*': {
					'*': ['*'],
					"": ["ast"],
				},
			},
		},
	}
	
	let output = JSON.parse(solcSnapshot.compile(JSON.stringify(input)));

	fs.writeFileSync("./code.json", JSON.stringify(output, null, 2));

	let targetContractName = "GraphToken";

	for (let fileName in output.contracts) {
		let subContracts = output.contracts[fileName];

		for (let contractName in subContracts) {
			if (contractName != targetContractName) {
				continue;
			}
	
			let contractData = subContracts[contractName];
			contracts.push({
				contractName,
				source,
				sourcePath: "./code.sol",
				ast: output.sources["code.sol"].ast,
				binary: `0x${contractData.evm.bytecode.object}`,
				sourceMap: contractData.evm.bytecode.sourceMap,
				deployedBinary: `0x${contractData.evm.deployedBytecode}`,
				deployedSourceMap: contractData.evm.deployedBytecode.sourceMap,
				abi: contractData.abi,
			});
		}
	}

	// fs.writeFileSync("./code.json", JSON.stringify(output, null, 2));

	console.log("compilation process complete.");
	
	promise_resolve();
});

	await promise;

	let transactionHash = "0x03e784bfcb8f6175f7a321586f671dff08a87070f62f5c9ae928e8f2c31e17bd";

	console.log("intiailizing debugger...");

	let bugger = await Debugger.forTx(
		transactionHash,
		{
			provider: web3.currentProvider,
			contracts,
		},
	);
	
	// console.log("connecting to debugger...");
	// let session = bugger.connect();
	// console.log("awaiting session ready...");
	// await session.ready();
	// console.log("debugger session ready!");
```

`code.sol`:

 - Go to https://etherscan.io/address/0xc944e90c64b2c07662a292be6244bdf05cda44a7
 - Go to the Contract tab and copy the contract code

`package.json`:

```
{
	"type": "module",
	"dependencies": {
		"@truffle/debugger": "^9.2.12",
		"ganache": "^7.0.0",
		"solc": "^0.8.11",
		"web3": "^1.7.0"
	}
}
```

You can get an API key from https://getblock.io -- just login with Google, then click "Production" in the sidebar of the console, then click on "generate endpoint", choose ethereum, and then choose mainnet.

**Comment by haltman-at** [comment]
OK, thanks, we'll take a look at this later!

> I actually was using Ganache's forking feature initially (after getting the email about it yesterday), but that was causing another error:

You should probably an issue [on the Ganche repo](https://github.com/trufflesuite/ganache/issues), then!

**Comment by haltman-at** [comment]
OK, so, the good news is I've put up a PR to fix this problem as #4661.

The bad news is, if I'm not mistaken, that the endpoint you're using doesn't support `debug_traceTransaction`, so you're still going to get an error.  But, that's a different matter.  Hope this is helpful (once it's released)!

**Comment by davidcallanan** [comment]
@haltman-at Thank you for everything! I'll try find a different endpoint :)

**Comment by haltman-at** [comment]
Great, glad it worked, thanks for bringing this problem to our attention!  Meanwhile the patch should go out officially later today. :)

**Comment by gnidan** [comment]
This is addressed and released in [Truffle v5.4.31](https://github.com/trufflesuite/truffle/releases/tag/v5.4.31). Thanks!
